src/Middleware/HeaderPropagation/src/HeaderPropagationMessageHandler.cs

```diff
@@ -0,0 +1,54 @@
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.Http.HeaderPropagation
+{
+    public class HeaderPropagationMessageHandler : DelegatingHandler
+    {
+        private readonly HeaderPropagationOptions _options;
+        private readonly IHttpContextAccessor _contextAccessor;
+
+        public HeaderPropagationMessageHandler(HeaderPropagationOptions options, IHttpContextAccessor contextAccessor)
+        {
+            _options = options;
+            _contextAccessor = contextAccessor;
+        }
+
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (_contextAccessor.HttpContext != null)
+            {
+                foreach (var header in _options.Headers)
+                {
+                    if (!_contextAccessor.HttpContext.Request.Headers.TryGetValue(header.InputName, out var values)
```

Comment:
This didn't fix the concurrency issue, these headers were supposed to be captured in an AspNetCore middleware.

Reply:
Sure, I'll implement the middleware.
Apologies if it wasn't clear in the description, I had just moved the PR as-is.